src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,960 @@
+    {
+      dx_mat(i, j) = dx(i, j, Location(i, j));
+      dy_mat(i, j) = dy(i, j, Location(i, j));
+    }
+  }
+  Orientation = arma::atan(dy_mat / dx_mat);
+  Orientation.transform( [](double val) { if(val < 0) return (val + arma::datum::pi);  else return (val);} );
+  
+  for(size_t j = 0; j < InImage.n_cols; ++j)
+  {
+    for(size_t i = 0; i < InImage.n_rows; ++i)
+    {
+      if( abs(dx_mat(i, j)) + abs(dy_mat(i, j)) < 1E-5)
+        Orientation(i, j) = 0.5 * arma::datum::pi;
+    }
+  }
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+Histogram(MatType const &Magnitude,
+              MatType const &Orientation, 
+              size_t downscale, size_t interp)
+{
+  //i don't think this function can be vectorized.
+
+  //n_orient: number of orientations per gradient scale
+  const size_t n_orient = this->options["n_orient"];
+  //size of HistArr: n_rbin * n_cbin * n_orient . . . (create in caller...)
+  const size_t n_rbin = (Magnitude.n_rows + downscale - 1) / downscale;
+  const size_t n_cbin = (Magnitude.n_cols + downscale - 1) / downscale;
+  double o_range, o;
+  o_range = arma::datum::pi / n_orient;
+
+  CubeType HistArr(n_rbin, n_cbin, n_orient);
+  HistArr.zeros();
+
+  size_t r, c, o1, o2;
+  for(size_t i = 0; i < Magnitude.n_rows; ++i)
+  {
+    for(size_t j = 0; j < Magnitude.n_cols; ++j)
+    {
+      r = i / downscale;
+      c = j / downscale;
+
+      if( interp != 0)
+      {
+        o = Orientation(i, j) / o_range;
+        o1 = ((size_t) o) % n_orient;
+        o2 = (o1 + 1) % n_orient;
+        HistArr(r, c, o1) += Magnitude(i, j) * (1 + (int)o - o);
+        HistArr(r, c, o2) += Magnitude(i, j) * (o - (int) o);
+      }
+      else
+      {
+        o1 = (size_t) (Orientation(i, j) / o_range + 0.5) % n_orient;
+        HistArr(r, c, o1) += Magnitude(i, j);
+      }
+    }
+  }
+
+  HistArr = HistArr / downscale;
+
+  for (size_t i = 0; i < HistArr.n_slices; ++i)
+    HistArr.slice(i) = arma::square(HistArr.slice(i));
+  
+  return HistArr;
+}
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+GetShrunkChannels(CubeType const &InImage, CubeType &reg_ch, CubeType &ss_ch)
+{
+  CubeType luv = this->RGB2LUV(InImage);
+  
+  const size_t shrink = this->options["shrink"];
+  const size_t n_orient = this->options["n_orient"];
+  const size_t grd_smooth_rad = this->options["grd_smooth_rad"];
+  const size_t grd_norm_rad = this->options["grd_norm_rad"];
+  const size_t num_channels = 13;
+  const size_t rsize = luv.n_rows / shrink;
+  const size_t csize = luv.n_cols / shrink;
+  CubeType channels(rsize, csize, num_channels);
+  
+  
+  size_t slice_idx = 0;
+  
+  for( slice_idx = 0; slice_idx < luv.n_slices; ++slice_idx)
+    channels.slice(slice_idx)
+    = this->bilinearInterpolation(luv.slice(slice_idx), (size_t)rsize, (size_t)csize);
+
+  double scale = 0.5;
+  
+  while(scale <= 1.0)
+  {
```

Comment:
Another picky comment, I would use a for loop here, that would make the number of iterations clearer.